flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/DynamicFilteringDependencyProcessor.java

```diff
@@ -52,50 +55,91 @@ public class DynamicFilteringDependencyProcessor implements ExecNodeGraphProcess
                     }
                 };
         execGraph.getRootNodes().forEach(node -> node.accept(dynamicFilteringScanCollector));
 
-        for (Map.Entry<BatchExecTableSourceScan, List<ExecNode<?>>> entry :
+        for (Map.Entry<BatchExecTableSourceScan, List<DescendantInfo>> entry :
                 dynamicFilteringScanDescendants.entrySet()) {
-            if (entry.getValue().size() == 1) {
-                ExecNode<?> next = entry.getValue().get(0);
-                if (next instanceof BatchExecMultipleInput) {
-                    // the source can be chained with BatchExecMultipleInput
-                    continue;
-                }
-            }
-            // otherwise we need dependencies
-            entry.getKey().setNeedDynamicFilteringDependency(true);
+            BatchExecTableSourceScan tableSourceScan = entry.getKey();
+            BatchExecDynamicFilteringDataCollector dynamicFilteringDataCollector =
+                    getDynamicFilteringDataCollector(tableSourceScan);
+
+            // Add exchange between collector and enforcer
+            BatchExecExchange exchange =
+                    new BatchExecExchange(
+                            context.getPlanner().getTableConfig(),
+                            InputProperty.builder()
+                                    .requiredDistribution(InputProperty.ANY_DISTRIBUTION)
+                                    .damBehavior(InputProperty.DamBehavior.BLOCKING)
+                                    .build(),
+                            (RowType) dynamicFilteringDataCollector.getOutputType(),
+                            "Exchange");
+            exchange.setRequiredExchangeMode(StreamExchangeMode.BATCH);
+            exchange.setInputEdges(
+                    Collections.singletonList(
+                            ExecEdge.builder()
+                                    .source(dynamicFilteringDataCollector)
+                                    .target(exchange)
+                                    .build()));
+
+            // set enforcer inputs
+            BatchExecExecutionOrderEnforcer enforcer =
+                    new BatchExecExecutionOrderEnforcer(
+                            context.getPlanner().getTableConfig(),
+                            Arrays.asList(
+                                    exchange.getInputProperties().get(0), InputProperty.DEFAULT),
+                            tableSourceScan.getOutputType(),
+                            "OrderEnforcer");
+            ExecEdge edge1 = ExecEdge.builder().source(exchange).target(enforcer).build();
+            ExecEdge edge2 = ExecEdge.builder().source(tableSourceScan).target(enforcer).build();
+            enforcer.setInputEdges(Arrays.asList(edge1, edge2));
+
+            // we clear the input of tableSourceScan to avoid cycle in exec plan
+            tableSourceScan.setInputEdges(Collections.emptyList());
+            tableSourceScan.setInputProperties(Collections.emptyList());
```

Comment:
Can you explain why we need to reset the `inputProperties`?  I'm concerned that the `inputProperties` is final from the initial design, we shouldn't change it as soon as possible. I think we don't need to reset it without having side effects, there are many places that generate the new Exchange without resetting it.

Reply:
In [InputPriorityGraphGenerator#updateTopologyGraph](https://github.com/apache/flink/blob/de01f021bdfead67459800abca6bbca9fcf3cecd/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/utils/InputPriorityGraphGenerator.java#L141C22-L141C35), it will check the size of `inputProperties` should be equal to the size of `inputEdges`.

Reply:
I think it would be better if we new a new TableScan here instead of changing the final `inputProperties `.

Reply:
Fixed